### PR TITLE
:sparkles: feat: Add support for RebuildTree

### LINKED ISSUE #2769: 🚀 [Feature]: Allow rebuilding route tree stack
### Feature Description

I was thinking it'd be awesome if the `buildTree` method in `App` could be made public. This tweak would open up possibilities for dynamic request handling.

already checked https://github.com/gofiber/fiber/issues/735 I get that rebuilding the tree might be resource-intensive, but we'd only use it during development. This way, we can ensure our routing behaves as expected without resorting to an `*` route and redoing all the routing and parsing logic.

Our goal is to enable users to define routes dynamically, with handlers loaded in on-the-fly. So, by simply making the **buildTree** method public, it'd really help us out. 

### Additional Context (optional)

The code snippet below illustrates how making `buildTree` public could streamline fiber routing without the need for using *.

```go
s.app.Get("/define", func(ctx *fiber.Ctx) error {
		// load config
		path := "/test/:param"
		s.app.Get(path, func(ctx *fiber.Ctx) error {
			// load the handler instruction
			// executing logic written by user
			// can access the :param easily
			return nil
		})
		// after each definition
		s.app.BuildTree()

		return ctx.JSON(map[string]any{
			"success": true,
		})
	})
```

### Code Snippet (optional)

_No response_

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my suggestion prior to opening this one.
- [X] I understand that improperly formatted feature requests may be closed without explanation.

### PR DESCRIPTION
**Pull request by luk3skyw4lker**
# Description

This PR adds the RebuildTree method, which allows for dynamic route registration. It also adds a lot of warnings and comments around the method because it is currently not thread-safe and very performance-intensive.

Fixes #2769 

## Changes introduced

A new RebuildTree public method was added in the App struct list of methods. It allows the user to dynamically register routes and call it to make the routes available for the server to handle. Tests were added and a warning in the method comment and in the docs about the not implemented thread-safety and performance of the method.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist

Before you submit your pull request, please make sure you meet these requirements:

- [x] Followed the inspiration of the Express.js framework for new functionalities, making them similar in usage.
- [x] Conducted a self-review of the code and provided comments for complex or critical parts.
- [x] Updated the documentation in the `/docs/` directory for [Fiber's documentation](https://docs.gofiber.io/).
- [x] Added or updated unit tests to validate the effectiveness of the changes or new features.
- [x] Ensured that new and existing unit tests pass locally with the changes.
- [x] Verified that any new dependencies are essential and have been agreed upon by the maintainers/community.
- [x] Aimed for optimal performance with minimal allocations in the new code.
- [x] Provided benchmarks for the new code to analyze and improve upon.

### CODE REVIEW

File: docs/api/app.md [hunk 1]
```diff
@@ -573,3 +573,13 @@
...
+
+## RebuildTree
+
+RebuildTree is a method destined to rebuild the route tree stack and allow dynamically route registers. It returns an app pointer.
+
+```go title="Signature"
+func (app *App) RebuildTree() *App
+```
+
+**NOTE**: This method should be used in the most careful way possible, since it's not currently possible to make it thread-safe (it would add a big performance overhead to do so) and calling it is very performance-intensive, so it's recommended to be used only in development mode and never concurrently.
```
[the remark this answers is not in the document]
Comment (luk3skyw4lker) [reply docs/api/app.md:585]: There is no doc for dynamic routing and no plans to add one.

File: router_test.go [hunk 2]
```diff
@@ -368,6 +369,33 @@ func Test_Router_NotFound_HTML_Inject(t *testing.T) {
...
+	require.Equal(t, http.StatusNotFound, resp.StatusCode, "Status code")
+
+	resp, err = app.Test(httptest.NewRequest(MethodGet, "/test", nil))
+	require.NoError(t, err, "app.Test(req)")
+	require.Equal(t, http.StatusOK, resp.StatusCode, "Status code")
+
+	resp, err = app.Test(httptest.NewRequest(MethodGet, "/dynamically-defined", nil))
+	require.NoError(t, err, "app.Test(req)")
+	require.Equal(t, http.StatusOK, resp.StatusCode, "Status code")
+}
```
[the remark this answers is not in the document]
Comment (luk3skyw4lker) [reply router_test.go:397]: There's not a short-term plan to support concurrency on this method yet.

File: router.go [hunk 3]
```diff
@@ -409,6 +409,21 @@ func (app *App) addRoute(method string, route *Route, isMounted ...bool) {
...
+// is only done in the startupProcess of the app, this method does not makes sure that the
+// routeTree is being safely changed, as it would add a great deal of overhead in the request.
+// Latest benchmark results showed a degradation from 82.79 ns/op to 94.48 ns/op and can be found in:
+// https://github.com/gofiber/fiber/issues/2769#issuecomment-2227385283
+func (app *App) RebuildTree() *App {
+	app.mutex.Lock()
+	defer app.mutex.Unlock()
+
+	return app.buildTree()
+}
```
[the remark this answers is not in the document]
Comment (luk3skyw4lker) [reply router.go:425]: There's not a short-term plan to support concurrency on this method yet.

**Comment by gaby** [comment]
@luk3skyw4lker The markdown can be fixed using this:

```console
markdownlint-cli2 "**/*.md" "#vendor"
```

**Comment by luk3skyw4lker** [comment]
@gaby thanks! I was crazy about it already. I'll fix it in a bit

**Comment by gaby** [comment]
@luk3skyw4lker Don't worry, CodeRabbit fixed it :-)

**Review by gaby — APPROVED** [review]
LGTM

File: docs/api/app.md [hunk 4]
```diff
@@ -573,3 +573,31 @@ Hooks is a method to return [hooks](./hooks.md) property.
 ```go title="Signature"
 func (app *App) Hooks() *Hooks
 ```
+
+## RebuildTree
```
Comment (efectn) [inline docs/api/app.md:577]: Can you update what's new guide too

**Review by efectn — COMMENTED** [review]
It may worth to add mutex to `addRoute`. it's not much matter for performance

File: docs/whats_new.md [hunk 5]
```diff
@@ -389,6 +389,31 @@
...
+
+    return c.SendStatus(http.StatusOK)
+})
+```
+
+In this example, a new route is defined and then `RebuildTree()` is called to make sure the new route is registered and available.
+
+**Note:** Use this method with caution. It is **not** thread-safe and calling it can be very performance-intensive, so it should be used sparingly and only in 
+development mode. Avoid using it concurrently.
+
```
[the remark this answers is not in the document]
Comment (luk3skyw4lker) [reply docs/whats_new.md]: There's a more detailed version with all the warnings also in the app.md file and this part references it.

File: docs/api/app.md (hunk 4 above)
Comment (luk3skyw4lker) [reply docs/api/app.md:577]: I think I have updated it in the right way. It's worth the review tho

**Comment by luk3skyw4lker** [comment]
@efectn adding the mutex in the addRoute function caused a deadlock. I guess that somehere before the mutex is already locked. And for this specific case I don't think it's needed because the buildTree function never calls addRoute or register, this is called somewhere else.

**Comment by efectn** [comment]
Yes you need to remove mutex here https://github.com/gofiber/fiber/blob/main/router.go#L403

**Comment by luk3skyw4lker** [comment]
My bad @efectn, that slipped right through me, sorry!

**Review by ReneWerner87 — APPROVED** [review]

**Comment by ReneWerner87** [comment]
@gaby @efectn can i merge or is there something missing ?

**Review by efectn — APPROVED** [review]

**Review by gaby — APPROVED** [review]
LGTM

**Pull request merged**